polkadot/node/network/collator-protocol/src/validator_side/claim_queue_state.rs

```diff
@@ -84,16 +97,42 @@ struct ClaimInfo {
 /// have got one claim which won't be satisfied in the worst case scenario.
 pub(crate) struct ClaimQueueState {
 	block_state: VecDeque<ClaimInfo>,
+	/// Represents blocks which are yet to be created. The claim queue at a leaf tells us what's
+	/// scheduled on the next two blocks. Since they are not yet authored we keep them separately
+	/// here.
 	future_blocks: VecDeque<ClaimInfo>,
+	/// Candidates with claimed slots per relay parent. We need this information in order to remove
+	/// claims. The key is the relay parent of the candidate and the value - the set of candidates
+	/// at it.
+	///
+	/// Note 1: We can't map the candidates to an exact slot since we need to keep track on their
+	/// ordering which will be an overkill in the context of `ClaimQueueState`. That's why we keep
+	/// the claims on first came first claimed basis.
+	///
+	/// Let's say there are three candidates built on top of each other A->B->C. We claim slots for
+	/// them in the order we receive them. If we receive them out of order the claims will be out
+	/// of order too, but this is fine since we only care about the count. If we fetch B and C
+	/// successfully and can't fetch A the user of `ClaimQueueState` must make sure that B and C
+	/// are removed.
+	///
+	/// Note 2: During pruning we remove all the candidates for the pruned relay parent because we
+	/// no longer need to know about them. If the claim was not undone so far - it will be
+	/// permanent.
+	candidates: HashMap<Hash, HashSet<CandidateHash>>,
```

Comment:
This hashmap adds some complexity and it only brings improvements if the queue has a lot of elements. Probably thousands. So just wanted to check if we actually need it ? What is the order of magnitude of the number of claims in the queue usually.

Reply:
For sure give it a try. Just keep in mind the forks and the (lack of) candidate ordering.